toolkit/tools/internal/grub/grubtokenizer.go

```diff
@@ -0,0 +1,509 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// Parses a grub config file into tokens.
+// This handles proper interpretation of metacharacters, newlines, variable expansions, character escapes, single quoted
+// strings, and double quoted strings. While this is not a full parser into an AST (abstract syntax tree), it is better
+// than parsing the grub config files using regexes.
+
+package grub
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/grub/filescanner"
+)
+
+func TokenizeGrubConfigFile(name string) ([]Token, error) {
+	configContent, err := os.ReadFile(name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read grub config file:\n%w", err)
+	}
+
+	return TokenizeGrubConfig(string(configContent))
+}
+
+func TokenizeGrubConfig(configContent string) ([]Token, error) {
+	scanner := filescanner.NewFileScanner(configContent)
+	tokenizer := grubConfigTokenizer{
+		scanner: scanner,
+	}
+
+	err := tokenizer.tokenize()
+	if err != nil {
+		return tokenizer.tokens, fmt.Errorf("failed to parse (tokenize) grub config file:\n%w", err)
+	}
+
+	return tokenizer.tokens, nil
+}
+
+type grubConfigTokenizer struct {
+	scanner  *filescanner.FileScanner
+	tokens   []Token
+	subWords []SubWord
+}
+
+func (t *grubConfigTokenizer) tokenize() error {
+	for {
+		c, eof := t.scanner.Peek()
+		if eof {
+			break
+		}
+
+		switch c {
+		// Metacharacters
+		case '{', '}', '|', '&', ';', '<', '>', '\n':
+			locStart := t.scanner.Loc()
+			// Consume metacharacter
+			t.scanner.Next()
+
+			var tokenType TokenType
+			switch c {
+			case '{':
+				tokenType = LBRACE
+			case '}':
+				tokenType = RBRACE
+			case '|':
+				tokenType = BAR
+			case '&':
+				tokenType = AND
+			case ';':
+				tokenType = SEMICOLON
+			case '<':
+				tokenType = LT
+			case '>':
+				tokenType = GT
+			case '\n':
+				tokenType = NEWLINE
+			}
+
+			token := t.newToken(locStart, t.scanner.Loc(), tokenType)
+			t.tokens = append(t.tokens, token)
+
+		// Space
+		case ' ', '\t':
+			err := t.parseSpace()
+			if err != nil {
+				return err
+			}
+
+		default:
+			err := t.parseWord()
+			if err != nil {
+				return err
+			}
+		}
+
+	}
+
+	return nil
+}
+
+func (t *grubConfigTokenizer) parseWord() error {
+	locStart := t.scanner.Loc()
+
+	c, eof := t.scanner.Peek()
+	if eof {
+		return nil
+	}
+
+	// Check if the word starts with '#'.
+	// Note: A '#' in the middle of a word doesn't start a comment.
+	if c == '#' {
+		err := t.parseComment()
+		if err != nil {
+			return err
+		}
+
+		return nil
```

Comment:
Nit-pick: the check seems redundant in this case - might as well just return `err`.
```suggestion
		return t.parseComment()
```